### DISCUSSION TITLE
Determination and assigning of MSS in TCP

### DISCUSSION THREAD

**Discussion by kisai19980124** [esnet/iperf#1440] [Q&A]
I am new to iPerf3 and recently checking the speed of the mobile network between my MacBook and an iPerf3 server.
I am curious about how MSS is determined if `-M` argument is not assigned. Is the MSS taking from the system?

BTW I am always getting "`iperf3: error - unable to set TCP/SCTP MSS: invalid argument`" error on my MacBook when assigning the MSS using "`-M 1448`". Does someone know how it is happening and is there any way to solve it?

**Comment by davidBar-On** [comment]
> I am curious about how MSS is determined if -M argument is not assigned. Is the MSS taking from the system?

The TCP stack calculates the default MSS size.  See for example [this IBM explanation](https://www.ibm.com/support/pages/what-mss-maximum-segment-size-and-how-it-calculated).

> I am always getting "`iperf3: error - unable to set TCP/SCTP MSS: invalid argument`" error on my MacBook when assigning the MSS using "`-M 1448`".

On my Linux `-M 1448` works o.k.  What is the reported default "Control connection MSS" MSS when you run the client with `-V`?

**Reply by kisai19980124** [reply]
Thank you for your replying.

For the first question, thank you for the reference. Is this a standard function included in the TCP/IP protocol?
Sorry for my poor network knowledge.

For the second question,
> What is the reported default "Control connection MSS" MSS when you run the client with `-V`?

The default MSS is 1448 if I didn't add MSS argument.
`TCP MSS: 1448 (default)`

Maybe it's the same problem as https://github.com/esnet/iperf/issues/779#issuecomment-411856679 ?

>`iperf3: error - unable to set TCP/SCTP MSS: Invalid argument`
I've reproduced this problem as well, although interestingly I see it under IPv4 but not IPv6 in my testing thus far.

Though I couldn't find further information about this issue.

**Comment by davidBar-On** [comment]
> Is this a standard function included in the TCP/IP protocol?

TCP SYN/ACK 3-way handshake is a standard TCP functionality.  The MSS calculation depends on the specific TCP stack implementation, although I assume that at least most of the implementations does something similar to the described calculation.

> Maybe it's the same problem as https://github.com/esnet/iperf/issues/779#issuecomment-411856679 ?

There is a good chance that it is, and there is there is a bug in the MacOs implementation.  Maybe someone else have further indite into this.
